File: 2001~2500/2390. Removing Stars From a String/main4.py
"""
Source: https://leetcode.com/problems/removing-stars-from-a-string/
Date: 2023/4/11
Skill:
Ref:
Runtime: 224 ms, faster than 77.42%
Memory Usage: 15.7 MB, less than 40.9%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools


class Solution:
    def removeStars(self, s: str) -> str:
        res = []
        for c in s:
            if c == "*":
                res.pop()
            else:
                res.append(c)

        return "".join(res)


if __name__ == "__main__":
    s = Solution()

File: 0501~1000/0856. Score of Parentheses/main4.cpp
/**
 * Source: is.gd/1fOV69
 * Date: 2023/11/30
 * Skill: Divide and conquer
 * Ref:
 * Runtime: 2 ms, faster than 50.85% of C++ online submissions
 * Memory Usage: 6.29 MB, less than 99.60% of C++ online submissions
 * Time complexity:
 * Space complexity:
 * Constraints:
 *
 */


#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;


class Solution {
public:
    int scoreOfParentheses(string s) {
        int depth = 0, res = 0;
        char prev = ' ';
        for (auto &c: s) {
            depth += c == '(' ? 1 : -1;
            if (c == ')' && prev == '(') res += 1 << depth;
            prev = c;
        }
        return res;
    }
};